Sts1CobcSw/Firmware/EduProgramTransferThread.cpp

```diff
@@ -0,0 +1,159 @@
+#include <Sts1CobcSw/Firmware/TopicsAndSubscribers.hpp>
+#include <Sts1CobcSw/FramSections/FramLayout.hpp>
+#include <Sts1CobcSw/FramSections/PersistentVariables.hpp>
+#include <Sts1CobcSw/Outcome/Outcome.hpp>
+#include <Sts1CobcSw/RodosTime/RodosTime.hpp>
+#include <Sts1CobcSw/Utility/DebugPrint.hpp>
+#include <Sts1CobcSw/Vocabulary/ProgramId.hpp>
+#include <Sts1CobcSw/Vocabulary/Time.hpp>
+
+#include <strong_type/difference.hpp>
+#include <strong_type/type.hpp>
+
+#include <rodos_no_using_namespace.h>
+
+#include <charconv>
+#include <cstdint>
+#include <system_error>
+#include <utility>
+
+
+namespace sts1cobcsw
+{
+namespace
+{
+constexpr auto stackSize = 1500U;
+constexpr auto eduProgramTransferThreadStartDelay = 15 * s;
+constexpr auto checkForProgramsInterval = 5 * s;
+
+
+[[nodiscard]] auto ReadProgramId(fs::DirectoryInfo const & info) -> Result<ProgramId>;
+auto HandleError(ErrorCode error) -> void;
+auto RemoveProgram(fs::Path const & file) -> void;
+auto SendProgramsToEdu(fs::DirectoryIterator & fileIterator) -> void;
+
+
+class EduProgramTransferThread : public RODOS::StaticThread<stackSize>
+{
+public:
+    EduProgramTransferThread()
+        : StaticThread("EduProgramQueueThread", eduProgramTransferThreadPriority)
+    {}
+
+
+private:
+    void run() override
+    {
+        SuspendFor(totalStartupTestTimeout);  // Wait for the startup tests to complete
+        SuspendFor(eduProgramTransferThreadStartDelay);
+        DEBUG_PRINT("Starting EDU program transfer thread\n");
+        TIME_LOOP(0, value_of(checkForProgramsInterval))
+        {
+            auto eduIsAlive = false;
+            eduIsAliveBufferForProgramTransfer.get(eduIsAlive);
+            if(edu::ProgramsAreAvailableOnCobc() && eduIsAlive)
+            {
+                DEBUG_PRINT("Programs available for transfer to EDU\n");
+                auto makeIteratorResult = fs::MakeIterator(edu::programsDirectory);
+                if(makeIteratorResult.has_error())
+                {
+                    DEBUG_PRINT("Failed to read program directory with error: %s\n",
+                                ToCZString(makeIteratorResult.error()));
+                }
+                else
+                {
+                    SendProgramsToEdu(makeIteratorResult.value());
+                }
+            }
+            DEBUG_PRINT_STACK_USAGE();
+        }
+    }
+} eduProgramQueueThread;
+
+
+auto ReadProgramId(fs::DirectoryInfo const & info) -> Result<ProgramId>
+{
+    auto const fileAppendixLength = 4;  // .zip
+    if(info.type != fs::EntryType::file)
+    {
+        return ErrorCode::invalidParameter;
+    }
+    if(info.name.size() > fileAppendixLength)
+    {
+        std::uint32_t value = 0;
+        auto result = std::from_chars(
+            info.name.data(), info.name.data() + (info.name.size() - fileAppendixLength), value);
+        if(result.ec == std::errc{})
+        {
+            return sts1cobcsw::ProgramId(value);
+        }
+    }
```

Comment:
This is basically the inverse of `edu::BuildProgramFilePath(programId) -> fs::Path` so I think this should be extracted into a function like `auto GetProgramId(fs::Path filename) -> Result<ProgramId>` and moved to Edu.hpp/cpp. In Edu.cpp you should then make `nProgramIdDigits` a file-scoped constant, introduce a constant for the file extension and use those to get the program ID out of the filename.